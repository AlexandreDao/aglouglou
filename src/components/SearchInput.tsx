import { View, TextInput, StyleSheet } from 'react-native'
import React, { forwardRef, useImperativeHandle, useRef, useState } from 'react'
import { ACTIVE_COLOR, TAB_BAR_BACKGROUND_COLOR, TEXT_COLOR } from '@/constants/colors'
import { IconSymbol } from '@/components/IconSymbol'
import SpeechRecognitionModal from '@/components/SpeechRecognitionModal'
import { Pressable } from 'react-native-gesture-handler'

export interface SearchInputRef {
  focus: () => void
}

interface SearchInputProps {
  onSubmitEditing?: (value: string) => void
  onFocus?: () => void
  onBlur?: () => void
  searchQuery?: string
  setSearchQuery?: (value: string) => void
}

const styles = StyleSheet.create({
  clearBtnContainer: {
    alignItems: 'center',
    borderRadius: 14,
    height: 48,
    justifyContent: 'center',
    width: 48,
  },
  container: {
    alignItems: 'center',
    backgroundColor: TAB_BAR_BACKGROUND_COLOR,
    borderColor: ACTIVE_COLOR,
    borderRadius: 14,
    borderWidth: 2,
    flex: 1,
    flexDirection: 'row',
    marginVertical: 8,
  },
  iconContainer: {
    alignItems: 'center',
    borderColor: ACTIVE_COLOR,
    borderRadius: 24,
    borderWidth: 2,
    height: 48,
    justifyContent: 'center',
    width: 48,
  },
  input: {
    color: TEXT_COLOR,
    flex: 1,
    height: 48,
  },
  inputCleared: {
    paddingHorizontal: 8,
  },
  inputDirty: {
    paddingLeft: 8,
  },
  row: {
    alignItems: 'center',
    flexDirection: 'row',
    gap: 8,
    justifyContent: 'center',
  },
})

const SearchInput = forwardRef<SearchInputRef, SearchInputProps>(
  ({ onSubmitEditing, onFocus, onBlur, searchQuery, setSearchQuery }, ref) => {
    const inputRef = useRef<TextInput>(null)
    const isSearchQueryCleared = !searchQuery?.trim()
    const [isOpen, setIsOpen] = useState(false)

    useImperativeHandle(ref, () => ({
      focus: () => inputRef.current?.focus(),
    }))

    return (
      <View style={styles.row}>
        <View style={styles.container}>
          <TextInput
            testID="search-input"
            ref={inputRef}
            value={searchQuery}
            onFocus={onFocus}
            onBlur={onBlur}
            onSubmitEditing={(e) => {
              onSubmitEditing?.(e?.nativeEvent?.text)
            }}
            submitBehavior="blurAndSubmit"
            onChangeText={setSearchQuery}
            placeholder={'Try typing "Irish"'}
            placeholderTextColor={ACTIVE_COLOR}
            cursorColor={TEXT_COLOR}
            style={[styles.input, isSearchQueryCleared ? styles.inputCleared : styles.inputDirty]}
            enterKeyHint="search"
            inputMode="search"
            returnKeyType="search"
            maxLength={256}
            autoCapitalize="none"
            autoFocus
            enablesReturnKeyAutomatically
          />
          {!isSearchQueryCleared && (
            <Pressable
              style={styles.clearBtnContainer}
              onPress={() => {
                if (inputRef.current) {
                  setSearchQuery?.('')
                  inputRef.current.clear()
                }
              }}
            >
              <IconSymbol name="xmark.circle.fill" color={TEXT_COLOR} />
            </Pressable>
          )}
        </View>
        {isSearchQueryCleared && (
          <Pressable style={styles.iconContainer} onPress={() => setIsOpen(true)}>
            <IconSymbol name="mic" color={TEXT_COLOR} />
          </Pressable>
        )}
        {isOpen && (
          <SpeechRecognitionModal
            isOpen={isOpen}
            setIsOpen={setIsOpen}
            onResult={(value) => {
              setSearchQuery?.(value)
              onSubmitEditing?.(value)
            }}
          />
        )}
      </View>
    )
  }
)

SearchInput.displayName = 'SearchInput'

export default SearchInput
