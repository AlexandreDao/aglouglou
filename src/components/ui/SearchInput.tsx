import { View, TextInput, Pressable, StyleSheet, InteractionManager, Keyboard } from 'react-native'
import React, { useCallback, useRef } from 'react'
import { ACTIVE_COLOR, TAB_BAR_BACKGROUND_COLOR, TEXT_COLOR } from '@/constants/colors'
import { IconSymbol } from '@/components/ui/IconSymbol'
import { useFocusEffect } from 'expo-router'

interface SearchInputProps {
  onSubmitEditing?: (value: string) => void
  onFocus?: () => void
  onBlur?: () => void
  searchQuery?: string
  setSearchQuery?: (value: string) => void
}

const styles = StyleSheet.create({
  clearBtnContainer: {
    alignItems: 'center',
    borderRadius: 14,
    height: 48,
    justifyContent: 'center',
    position: 'absolute',
    right: 0,
    width: 48,
  },
  container: {
    alignItems: 'center',
    flexDirection: 'row',
    gap: 8,
  },
  iconContainer: {
    alignItems: 'center',
    borderColor: ACTIVE_COLOR,
    borderRadius: 24,
    borderWidth: 2,
    height: 48,
    justifyContent: 'center',
    width: 48,
  },
  input: {
    backgroundColor: TAB_BAR_BACKGROUND_COLOR,
    borderColor: ACTIVE_COLOR,
    borderRadius: 14,
    borderWidth: 2,
    color: TEXT_COLOR,
    flex: 1,
    height: 48,
    marginVertical: 8,
    paddingHorizontal: 8,
  },
})

function SearchInput({ onSubmitEditing, onFocus, onBlur, searchQuery, setSearchQuery }: SearchInputProps) {
  const inputRef = useRef<TextInput>(null)

  useFocusEffect(
    useCallback(() => {
      const timer = setTimeout(() => {
        inputRef.current?.focus()
      }, 100)
      return () => clearTimeout(timer)
    }, [])
  )

  return (
    <View style={styles.container}>
      <TextInput
        ref={inputRef}
        value={searchQuery}
        onFocus={onFocus}
        onBlur={onBlur}
        onSubmitEditing={(e) => {
          onSubmitEditing?.(e.nativeEvent.text)
        }}
        submitBehavior="blurAndSubmit"
        onChangeText={setSearchQuery}
        placeholder="Search"
        placeholderTextColor={ACTIVE_COLOR}
        cursorColor={TEXT_COLOR}
        style={styles.input}
        enterKeyHint="search"
        inputMode="search"
        maxLength={256}
        returnKeyType="search"
        selectTextOnFocus
      />
      {!searchQuery?.trim() ? (
        <Pressable style={styles.iconContainer}>
          <IconSymbol name="mic" color={TEXT_COLOR} />
        </Pressable>
      ) : (
        <Pressable
          style={styles.clearBtnContainer}
          onPress={() => {
            if (inputRef.current) {
              setSearchQuery?.('')
              inputRef.current.clear()
            }
          }}
        >
          <IconSymbol name="xmark.circle.fill" color={TEXT_COLOR} />
        </Pressable>
      )}
    </View>
  )
}

export default SearchInput
