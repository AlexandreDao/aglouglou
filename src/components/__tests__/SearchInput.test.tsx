import React from 'react'
import SearchInput from '@/components/SearchInput'
import { fireEvent, render } from '@testing-library/react-native'

describe('<SearchInput />', () => {
  test('SearchInput snapshot', () => {
    const tree = render(<SearchInput />).toJSON()

    expect(tree).toMatchSnapshot()
  })
  test('SearchInput searchQuery', () => {
    const searchQuery = 'Hello world'
    const { getByDisplayValue } = render(<SearchInput searchQuery={searchQuery} />)

    expect(getByDisplayValue(searchQuery)).toBeDefined()
  })
  test('SearchInput focus', () => {
    const onFocusMock = jest.fn()
    const { getByTestId } = render(<SearchInput onFocus={onFocusMock} />)

    fireEvent(getByTestId('search-input'), 'focus')
    expect(onFocusMock).toHaveBeenCalled()
  })
  test('SearchInput onSubmit', () => {
    const text = 'Hello world'
    const onSubmitMock = jest.fn()
    const { getByTestId } = render(<SearchInput onSubmitEditing={onSubmitMock} />)
    const input = getByTestId('search-input')

    fireEvent.changeText(input, text)
    fireEvent(input, 'submitEditing', { nativeEvent: { text } })
    expect(onSubmitMock).toHaveBeenCalled()
  })
})
