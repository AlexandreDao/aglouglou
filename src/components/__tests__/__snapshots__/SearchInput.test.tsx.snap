// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> SearchInput snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "gap": 8,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#380F1F",
        "borderColor": "#FF8EB6",
        "borderRadius": 14,
        "borderWidth": 2,
        "flex": 1,
        "flexDirection": "row",
        "marginVertical": 8,
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoFocus={true}
      cursorColor="#EDEEF0"
      enablesReturnKeyAutomatically={true}
      enterKeyHint="search"
      inputMode="search"
      maxLength={256}
      onSubmitEditing={[Function]}
      placeholder="Try typing "Irish""
      placeholderTextColor="#FF8EB6"
      returnKeyType="search"
      style={
        [
          {
            "color": "#EDEEF0",
            "flex": 1,
            "height": 48,
          },
          {
            "paddingHorizontal": 8,
          },
        ]
      }
      submitBehavior="blurAndSubmit"
      testID="search-input"
    />
  </View>
  <RNGestureHandlerButton
    collapsable={false}
    enabled={true}
    hitSlop={
      {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      }
    }
    rippleColor={0}
    style={
      [
        {},
        {
          "alignItems": "center",
          "borderColor": "#FF8EB6",
          "borderRadius": 24,
          "borderWidth": 2,
          "height": 48,
          "justifyContent": "center",
          "width": 48,
        },
      ]
    }
  >
    <ViewManagerAdapter_SymbolModule
      animated={false}
      colors={[]}
      name="mic"
      resizeMode="scaleAspectFit"
      style={
        [
          {
            "height": 24,
            "width": 24,
          },
          [
            {
              "height": 24,
              "width": 24,
            },
            undefined,
          ],
        ]
      }
      tintColor="#EDEEF0"
      type="monochrome"
      weight="regular"
    />
  </RNGestureHandlerButton>
</View>
`;
